rfcs/framework-0040-relations.md

```diff
@@ -0,0 +1,321 @@
+    )
+```
+
+Note that the class property name (``relations``) can be named whatever you like (e.g. ``vocabularies``).
+
+We proposed four relationship types:
+
+- ``PIDRelation`` - a relation using PIDStore for lookup.
+- ``PKRelation`` - a relation using record UUIDs for lookup (i.e. a record's primary key).
+- ``PIDListRelation`` - same as ``PIDRelation`` but for list of referenced objects.
+- ``PKListRelation`` - same as ``PKRelation`` but for list of objects.
+
+An alternative, would be to specify relations individually:
+
+```python
+class Record(RecordBase):
+    resource_type = PIDRelation('metadata.resource_type', ResourceTypeRecord.pid)
+```
+
+However, this will make it a bit more diffcult to operate on all relations at once (e.g. ``record.relations.dereference()``).
+
+##### Limiting attributes
+
+Limiting the attributes could be done as:
+
+```python
+PIDRelation('metadata.resource_type', ResourceTypeRecord.pid, attrs=[
+    'type', 'subtype', 'some.nested.key'])
+```
+
+#### Dereferencing / retrieving the record
+
+```python
+# No relation defined
+record.relations.resource_type() is None
+
+# Getting via metadata
+record['metadata']['resource_type']
+# - return only id if  not dereferenced
+# - return id, type and subtype
+
+# Get the related record:
+# - if not dereferenced: query db and store dereferenced
+# - if dereferenced: use dereferenced
+rt = record.relations.resource_type()
+# Get the related record (force a DB query)
+rt = record.relations.resource_type(force=True)
+
+# Get many records:
+for s in record.relations.subjects():
+    # ...
+
+# Dereference all records at once
+record.relations.dereference()
+```
+
+#### Setting related records
+
+
+```python
+# Via metadata
+Record({
+    "metadata": {
+        "resource_type": {"id": "text-article"}
+    }
+})
+
+# Via system field
+record.relations.resource_type = ResourceType.get_record('text-article')
+record.relations.resource_type = 'text-article'
+
+
+record.relations.subjects.append('3c1eb018-0031-4488-b11d-ad6831997e89')
+```
+
+
+#### Integrity checking and validation
+
+
+```python
+# Checking if a specific ID exists
+Record.relations.resource_type.exists('text-article')
+    # -> ResourceTypeRecord.pid.resolve('text-article')
+
+# Checking if many ids in one go.
+Record.relations.subjects.exists_many(['3c1eb018-0031-4488-b11d-ad6831997e89', ...])
+    # -> SubjectsRecord.get_record('3c1eb018-0031-4488-b11d-ad6831997e89'))
+
+# Validate all relations
+record.relations.validate()
+# Above could be hooked directly into "record.validate()"
+```
+
+#### Updating outdated dereferenced relations
+
+Updating the outdated references is not doable only via the data access layer.
+
+```python
+# API to scan for records
+service.scan(relations={'resource_type': [('text-article', 18), ...]})
+
+# API to using above scan to reindex
+service.reindex(relations={'resource_type': [('text-article', 18), ...]}, bulk=False)
+```
```

Comment:
These are just sugar (more convenient syntax), right? We already can search and reindex on ES indices, correct?